util/man/start.8
.TH start 8 "September 2006" "Upstart"
.\"
.SH NAME
start, stop, status \- job control utilities
.\"
.SH SYNOPSIS
\fBstart\fR [\fIOPTION\fR]... \fIJOB\fR...

\fBstop\fR [\fIOPTION\fR]... \fIJOB\fR...

\fBstatus\fR [\fIOPTION\fR]... \fIJOB\fR...
.\"
.SH DESCRIPTION
These utilities allow a system administrator to
.BR start ", " stop " or query the " status
of the jobs named on their command-line.
.\"
.SH AUTHOR
Written by Scott James Remnant.
.\"
.SH REPORTING BUGS
Report bugs at https://launchpad.net/products/upstart/+bugs
.\"
.SH COPYRIGHT
Copyright \(co 2006 Canonical Ltd.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.\"
.SH SEE ALSO
.BR init (8)


NEWS
0.3.0  2006-10-17

	* Reverted logd behaviour from previous version, it's up to the
	  init scripts to send messages to the console if they wish.

	* Compatibility programs must now be explicitly enabled by using
	  ./configure --enable-compat=sysv

	* "shutdown" and "reboot" are now considered System V compatibility
	  programs, as they emulate the behaviour of those.  Dropped some
	  added options to make them fit.

	* All programs given improved --help text.

	* "initctl" rewritten, any sub-command can be run directly by
	  making it a symlink to "initctl" itself.

	* "start", "stop" and "status" are now just symlinks to "initctl";
	  not a separate binary.

0.2.7  2006-09-20

	* logd writes received messages to the console unless "quiet" is
	  on the kernel command-line

	* runaway jobs are now caught when they start, rather than respawn,
	  so stop/start loops are caught

	* Include inotify support for compiling under glibc 2.3

0.2.6  2006-09-13

	* Fix major bug on architectures with 64-bit kernel and 32-bit
	  user-land caused by an inconsitency between the behaviour of
	  kernel's compat_sys_waitid() vs. sys_waitid() functions.

	* "halt" now only calls "shutdown -h now"

0.2.5  2006-09-09

	* "control-alt-delete" event name changed to "ctrlaltdel".

	* "initctl shutdown EVENT" added that performs the same job as
	  "shutdown" but without all the usual warnings, timings, etc.

	* "logd" has now been written, if installed this is started by init
	  before sending the "startup" call and all jobs with "console logged"
	  (the default) will have their output sent to this daemon.  It
	  currently just logs to /var/log/boot.

	* "shutdown -k" implemented.

	* The "shutdown" utility has been changed to generate "system-halt"
	  for "-H", "power-off" for "-P" and just "halt" if only "-h" given.

	* If "shutdown" is run when running under sysvinit, it will now
	  send the appropriate /dev/initctl message to allow upgrades.

	* "telinit S" implemented.

	* Instead of trying to start or stop jobs, "telinit" now just sends
	  "runlevel-X" events.

	* The "telinit" utility now ensures a "shutdown" event is sent
	  before switching to runlevel 0, 1 or 6.

	* If "telinit" is installed and init is called by the super-user,
	  "telinit" is invoked instead.

	* Basic manual pages included.

0.2.1  2006-09-01

	* Compilation fixes

0.2.0  2006-09-01

	* "shutdown", "reboot", "halt" and "poweroff" utilities provided
	  that match their traditional equivalents.

	* "start", "stop" and "status" utilities provided to start, stop
	  and query the status of jobs respectively.

	* "runlevel" and "telinit" utilities provided for compatibility.

	* "initctl list" will list active jobs.

	* Events vastly simplified to just simple strings.

	* Jobs now generate "jobname/start", "jobname/started",
	  "jobname/stop" and "jobname/stopped" events as they go through
	  state transitions.

	* Services generate a "jobname" event when they are running.

	* Tasks generate a "jobname" event when they have finished.

	* The "shutdown" utility will generate a "shutdown" event followed
	  by one of "maintenance", "reboot", "halt" or "poweroff" or any
	  admin-specified event.

	* "stalled" event generated when no jobs are running or queued.

	* "control-alt-delete" event generated when that key combination
	  is pressed

	* "kbdrequest" event generated when Alt-UpArrow is pressed

	* Runaway respawning services will now be caught.

	* init will re-exec on receipt of the SIGUSR1 signal.

0.1.1  2006-08-25

	* Minor bug fixes.

0.1.0  2006-08-24
 
	* Initial public release.



initctl/start.c
/* upstart
 *
 * Copyright © 2006 Canonical Ltd.
 * Author: Scott James Remnant <scott@ubuntu.com>.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
 */

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif /* HAVE_CONFIG_H */


#include <stdio.h>
#include <string.h>
#include <unistd.h>

#include <nih/macros.h>
#include <nih/alloc.h>
#include <nih/main.h>
#include <nih/option.h>
#include <nih/logging.h>
#include <nih/error.h>

#include <upstart/control.h>
#include <upstart/job.h>


/* Operation modes */
enum {
	START,
	STOP,
	STATUS
};


/**
 * options:
 *
 * Command-line options accepted.
 **/
static NihOption options[] = {
	NIH_OPTION_LAST
};


int
main (int   argc,
      char *argv[])
{
	char **args, **arg;
	int    mode, sock;

	nih_main_init (argv[0]);
	nih_option_set_usage ("JOB...");

	mode = START;
	if (! strcmp (program_name, "stop")) {
		mode = STOP;
	} else if (! strcmp (program_name, "status")) {
		mode = STATUS;
	}

	args = nih_option_parser (NULL, argc, argv, options, FALSE);
	if (! args)
		exit (1);

	/* Check we're root */
	setuid (geteuid ());
	if (getuid ()) {
		nih_error (_("Need to be root"));
		exit (1);
	}

	/* Connect to the daemon */
	sock = upstart_open ();
	if (sock < 0) {
		NihError *err;

		err = nih_error_get ();
		nih_error (_("Unable to establish control socket: %s"),
			   err->message);
		exit (1);
	}


	/* Iterate arguments */
	for (arg = args; *arg; arg++) {
		UpstartMsg msg, *reply;

		/* Build the message to send */
		switch (mode) {
		case START:
			msg.type = UPSTART_JOB_START;
			msg.job_start.name = *arg;
			break;
		case STOP:
			msg.type = UPSTART_JOB_STOP;
			msg.job_stop.name = *arg;
			break;
		case STATUS:
			msg.type = UPSTART_JOB_QUERY;
			msg.job_stop.name = *arg;
			break;
		}

		/* Send the message */
		if (upstart_send_msg (sock, &msg) < 0) {
			NihError *err;

			err = nih_error_get ();
			nih_error (_("Unable to send message: %s"),
				   err->message);
			exit (1);
		}

		/* Wait for the reply */
		reply = upstart_recv_msg (NULL, sock, NULL);
		if (! reply) {
			NihError *err;

			err = nih_error_get ();
			nih_error (_("Error receiving message: %s"),
				   err->message);
			exit (1);
		}

		switch (reply->type) {
		case UPSTART_JOB_STATUS:
			printf ("%s (%s) %s", reply->job_status.name,
				job_goal_name (reply->job_status.goal),
				job_state_name (reply->job_status.state));

			if (reply->job_status.state == JOB_WAITING) {
				printf ("\n");

			} else if ((reply->job_status.process_state == PROCESS_SPAWNED)
				   || (reply->job_status.process_state == PROCESS_NONE)) {
				printf (", process %s\n",
					process_state_name (reply->job_status.process_state));
			} else {
				printf (", process %d %s\n",
					reply->job_status.pid,
					process_state_name (reply->job_status.process_state));
			}
			break;
		case UPSTART_JOB_UNKNOWN:
			fprintf (stderr, _("%s: Unknown job: %s\n"),
				 program_name, reply->job_unknown.name);
			break;
		default:
			fprintf (stderr, _("%s: Unexpected reply (type %d)\n"),
				 program_name, reply->type);
		}

		nih_free (reply);
	}

	return 0;
}
